diff --git a/Makefile b/Makefile
index 593ea4e98..736fb6704 100644
--- a/Makefile
+++ b/Makefile
@@ -104,7 +104,7 @@ DEB_DEPENDS += python3-ply libunwind-dev
 DEB_DEPENDS += cmake ninja-build python3-jsonschema python3-yaml
 DEB_DEPENDS += python3-venv  # ensurepip
 DEB_DEPENDS += python3-dev python3-pip
-DEB_DEPENDS += libnl-3-dev libnl-route-3-dev libmnl-dev
+DEB_DEPENDS += libmnl-dev
 # DEB_DEPENDS += enchant  # for docs
 DEB_DEPENDS += python3-virtualenv
 DEB_DEPENDS += libssl-dev
@@ -112,7 +112,7 @@ DEB_DEPENDS += libelf-dev libpcap-dev # for libxdp (af_xdp)
 DEB_DEPENDS += iperf3 # for 'make test TEST=vcl'
 DEB_DEPENDS += iperf ethtool  # for 'make test TEST=vm_vpp_interfaces'
 DEB_DEPENDS += libpcap-dev
-DEB_DEPENDS += tshark
+#DEB_DEPENDS += tshark
 DEB_DEPENDS += jq # for extracting test summary from .json report (hs-test)
 DEB_DEPENDS += libiberty-dev
 DEB_DEPENDS += nasm libnuma-dev # for make-ext-deps
diff --git a/build/external/packages/dpdk.mk b/build/external/packages/dpdk.mk
index 57f8242e5..63220f0a9 100644
--- a/build/external/packages/dpdk.mk
+++ b/build/external/packages/dpdk.mk
@@ -17,7 +17,7 @@ DPDK_DEBUG                   ?= n
 DPDK_TAP_PMD                 ?= n
 DPDK_FAILSAFE_PMD            ?= n
 DPDK_MACHINE                 ?= generic
-DPDK_MLX_IBV_LINK            ?= static
+DPDK_MLX_IBV_LINK            ?= shared
 # On most of the systems, default value for max lcores is 128
 DPDK_MAX_LCORES              ?=
 
diff --git a/build/external/packages/xdp-tools.mk b/build/external/packages/xdp-tools.mk
index a692aacf1..0265b8eaf 100644
--- a/build/external/packages/xdp-tools.mk
+++ b/build/external/packages/xdp-tools.mk
@@ -25,7 +25,7 @@ define  xdp-tools_config_cmds
 endef
 
 define  xdp-tools_build_cmds
-	@cd ${xdp-tools_src_dir} && $(MAKE) CC=gcc V=1 BUILD_STATIC_ONLY=y PREFIX='$(xdp-tools_install_dir)' > $(xdp-tools_build_log)
+	@cd ${xdp-tools_src_dir} && $(MAKE) CC=gcc FORCE_SUBDIR_LIBBPF=1 V=1 BUILD_STATIC_ONLY=y PREFIX='$(xdp-tools_install_dir)' > $(xdp-tools_build_log)
 endef
 
 define  xdp-tools_install_cmds
diff --git a/src/plugins/vxlan/vxlan.c b/src/plugins/vxlan/vxlan.c
index 0885550d2..8b8cd66e4 100644
--- a/src/plugins/vxlan/vxlan.c
+++ b/src/plugins/vxlan/vxlan.c
@@ -43,7 +43,7 @@
  */
 
 
-vxlan_main_t vxlan_main;
+__clib_export vxlan_main_t vxlan_main;
 
 static u32
 vxlan_eth_flag_change (vnet_main_t *vnm, vnet_hw_interface_t *hi, u32 flags)
diff --git a/src/scripts/version b/src/scripts/version
index b8748af15..8f88c53c7 100755
--- a/src/scripts/version
+++ b/src/scripts/version
@@ -17,6 +17,12 @@ path=$( cd "$(dirname "${BASH_SOURCE}")" ; pwd -P )
 
 cd "$path"
 
+# Check if CUSTOM_VERSION is set
+if [ -n "$CUSTOM_VERSION" ]; then
+    echo "$CUSTOM_VERSION"
+    exit 0
+fi
+
 if [ -f .version ]; then
     vstring=$(cat .version)
 else
