diff --git a/src/plugins/sflow/node.c b/src/plugins/sflow/node.c
index 3d5363166..461555a2e 100644
--- a/src/plugins/sflow/node.c
+++ b/src/plugins/sflow/node.c
@@ -73,26 +73,30 @@ sflow_node_ingress_egress (vlib_main_t *vm, vlib_node_runtime_t *node,
   from = vlib_frame_vector_args (frame);
   n_left_from = frame->n_vectors;
 
+  u32 dir = (sample_type == SFLOW_SAMPLETYPE_EGRESS) ? VLIB_TX : VLIB_RX;
+  u32 sw_if = vnet_buffer (vlib_get_buffer (vm, from[0]))->sw_if_index[dir];
+  vnet_hw_interface_t *hw = vnet_get_sup_hw_interface(smp->vnet_main, sw_if);
+  sflow_per_interface_data_t *sfif = vec_elt_at_index(smp->per_interface_data, hw->hw_if_index);
+
   uword thread_index = os_get_thread_index ();
   sflow_per_thread_data_t *sfwk =
     vec_elt_at_index (smp->per_thread_data, thread_index);
 
-  /* note that sfwk->skip==1 means "take the next packet",
-     so we never see sfwk->skip==0. */
-
+  /* note that sfif->skip==1 means "take the next packet",
+     so we never see sfif->skip==0. */
   u32 pkts = n_left_from;
-  if (PREDICT_TRUE (sfwk->skip > pkts))
+  if (PREDICT_TRUE (sfif->skip > pkts))
     {
       /* skip the whole frame-vector */
-      sfwk->skip -= pkts;
-      sfwk->pool += pkts;
+      sfif->skip -= pkts;
+      sfif->pool += pkts;
     }
   else
     {
-      while (pkts >= sfwk->skip)
+      while (pkts >= sfif->skip)
 	{
 	  /* reach in to get the one we want. */
-	  vlib_buffer_t *bN = vlib_get_buffer (vm, from[sfwk->skip - 1]);
+	  vlib_buffer_t *bN = vlib_get_buffer (vm, from[sfif->skip - 1]);
 
 	  /* Sample this packet header. */
 	  u32 hdr = bN->current_length;
@@ -123,7 +127,7 @@ sflow_node_ingress_egress (vlib_main_t *vm, vlib_node_runtime_t *node,
 
 	  sflow_sample_t sample = {
 	    .sample_type = sample_type,
-	    .samplingN = sfwk->smpN,
+	    .samplingN = sfif->samplingN ? sfif->samplingN : smp->samplingN,
 	    .input_if_index = if_index,
 	    .output_if_index = if_index_out,
 	    .sampled_packet_size =
@@ -149,16 +153,16 @@ sflow_node_ingress_egress (vlib_main_t *vm, vlib_node_runtime_t *node,
 	  sfwk->smpl++;
 	  memcpy (sample.header, en, hdr);
 	  if (PREDICT_FALSE (!sflow_fifo_enqueue (&sfwk->fifo, &sample)))
-	    sfwk->drop++;
+	  sfwk->drop++;
 
-	  pkts -= sfwk->skip;
-	  sfwk->pool += sfwk->skip;
-	  sfwk->skip = sflow_next_random_skip (sfwk);
+	  pkts -= sfif->skip;
+	  sfif->pool += sfif->skip;
+	  sfif->skip = sflow_next_random_skip_if(smp, sfif);
 	}
       /* We took a sample (or several) from this frame-vector, but now we are
 	 skipping the rest. */
-      sfwk->skip -= pkts;
-      sfwk->pool += pkts;
+	  sfif->skip -= pkts;
+      sfif->pool += pkts;
     }
 
   /* the rest of this is boilerplate code just to make sure
diff --git a/src/plugins/sflow/sflow.c b/src/plugins/sflow/sflow.c
index 67b0cb38b..f231fb524 100644
--- a/src/plugins/sflow/sflow.c
+++ b/src/plugins/sflow/sflow.c
@@ -550,12 +550,18 @@ read_node_counters (sflow_main_t *smp, sflow_err_ctrs_t *ctrs)
     {
       sflow_per_thread_data_t *sfwk =
 	vec_elt_at_index (smp->per_thread_data, thread_index);
-      ctrs->counters[SFLOW_ERROR_PROCESSED] += sfwk->pool;
       ctrs->counters[SFLOW_ERROR_SAMPLED] += sfwk->smpl;
       ctrs->counters[SFLOW_ERROR_DROPPED] += sfwk->drop;
       ctrs->counters[SFLOW_ERROR_DIPROCESSED] += sfwk->dsmp;
       ctrs->counters[SFLOW_ERROR_DIDROPPED] += sfwk->ddrp;
     }
+
+  for (int i = 0; i < vec_len (smp->per_interface_data); i++)
+  {
+    sflow_per_interface_data_t *sfif = vec_elt_at_index (smp->per_interface_data, i);
+    if (sfif)
+      ctrs->counters[SFLOW_ERROR_PROCESSED] += sfif->pool;
+  }
 }
 
 static void
@@ -676,6 +682,19 @@ sflow_set_worker_sampling_state (sflow_main_t *smp)
     }
 }
 
+static void
+sflow_set_interface_sampling_rate (sflow_main_t *smp)
+{
+  for (int i = 0; i < vec_len(smp->per_interface_data); i++)
+  {
+    sflow_per_interface_data_t *sfif = vec_elt_at_index(smp->per_interface_data, i);
+    if (sfif && sfif->sflow_enabled && !sfif->samplingN)
+    {
+      sfif->skip = sflow_next_random_skip_if(smp,sfif);
+    }
+  }
+}
+
 static void
 sflow_sampling_start (sflow_main_t *smp)
 {
@@ -748,6 +767,7 @@ sflow_sampling_rate (sflow_main_t *smp, u32 samplingN)
       // dynamic change of sampling rate
       smp->samplingN = samplingN;
       sflow_set_worker_sampling_state (smp);
+      sflow_set_interface_sampling_rate(smp);
     }
   else
     {
@@ -889,14 +909,47 @@ sflow_enable_disable (sflow_main_t *smp, u32 sw_if_index, bool enable_disable)
       sfif->hw_if_index = sw->hw_if_index;
       sfif->polled = 0;
       sfif->sflow_enabled = enable_disable;
+
+      if(enable_disable)
+        {
+          sfif->seed = sw->hw_if_index;
+          sfif->skip = sflow_next_random_skip_if(smp,sfif);
+        }  
+
       sflow_enable_disable_interface (smp, sfif);
       smp->interfacesEnabled += (enable_disable) ? 1 : -1;
     }
-
   sflow_sampling_start_stop (smp);
   return 0;
 }
 
+int
+sflow_set_sampling_rate_set(sflow_main_t *smp, u32 sw_if_index, u32 samplingN)
+{
+  vnet_sw_interface_t *sw;
+
+  /* Utterly wrong? */
+  if (pool_is_free_index (smp->vnet_main->interface_main.sw_interfaces,
+    sw_if_index))
+  return VNET_API_ERROR_INVALID_SW_IF_INDEX;
+
+  /* Not a physical port? */
+  sw = vnet_get_sw_interface (smp->vnet_main, sw_if_index);
+  if (sw->type != VNET_SW_INTERFACE_TYPE_HARDWARE)
+    return VNET_API_ERROR_INVALID_SW_IF_INDEX;
+
+  SFLOW_DBG ("sw_if_index=%u, sup_sw_if_index=%u, hw_if_index=%u\n", sw->sw_if_index, sw->sup_sw_if_index, sw->hw_if_index);
+
+  vec_validate (smp->per_interface_data, sw->hw_if_index);
+  sflow_per_interface_data_t *sfif = vec_elt_at_index (smp->per_interface_data, sw->hw_if_index);
+
+  sfif->samplingN = samplingN;
+  sfif->skip = sflow_next_random_skip_if(smp, sfif);
+
+  return 0; 
+
+}
+
 int 
 sflow_report_linux_ifindex(sflow_main_t *smp, bool enable)
 {
diff --git a/src/plugins/sflow/sflow.h b/src/plugins/sflow/sflow.h
index 0540cae48..00ad8051e 100644
--- a/src/plugins/sflow/sflow.h
+++ b/src/plugins/sflow/sflow.h
@@ -247,4 +247,14 @@ sflow_next_random_skip (sflow_per_thread_data_t *sfwk)
   return (random_u32 (&sfwk->seed) % lim) + 1;
 }
 
+static inline u32
+sflow_next_random_skip_if(sflow_main_t *smp, sflow_per_interface_data_t *sfif)
+{
+  u32 effectiveN = sfif->samplingN ? sfif->samplingN : smp->samplingN;
+  if(effectiveN <= 1)
+    return 1;
+  u32 lim = (2 * effectiveN) - 1;
+  return (random_u32(&sfif->seed) % lim) + 1;
+}
+
 #endif /* __included_sflow_h__ */
diff --git a/src/plugins/sflow/sflow_common.h b/src/plugins/sflow/sflow_common.h
index 3e1021e4a..d699ceac0 100644
--- a/src/plugins/sflow/sflow_common.h
+++ b/src/plugins/sflow/sflow_common.h
@@ -19,6 +19,10 @@ typedef struct
   u32 linux_if_index;
   u32 polled;
   int sflow_enabled;
+  u32 samplingN;
+  u32 skip;
+  u32 pool;
+  u32 seed;
 } sflow_per_interface_data_t;
 
 /* mirror sflow_direction enum in sflow.api */
