diff --git a/configure.ac b/configure.ac
index 43fe79fa..5c55ac9d 100644
--- a/configure.ac
+++ b/configure.ac
@@ -34,12 +34,12 @@ AM_COND_IF([DEBUG],
             AC_MSG_ERROR("Flag conflict --enable-debug 'DEBUG' and --enable-code-coverage 'NDEBUG'"))])
 
 AC_ARG_WITH(sai,
-[  --with-sai=vs           compiling against specific SAI],
+[  --with-sai=vpp           compiling against specific SAI],
 [case "${withval}" in
-	vs)   sai=vs ;;
+	vpp)   sai=vpp ;;
 	*) AC_MSG_ERROR(bad value ${withval} for --with-sai) ;;
 esac])
-AM_CONDITIONAL(SAIVS,   test x$sai = xvs)
+AM_CONDITIONAL(SAIVPP,   test x$sai = xvpp)
 
 AC_ARG_WITH(dashsai,
     [ --with-dashsai         compiling against DASH SAI],
@@ -68,7 +68,7 @@ esac],[syncd=true])
 AS_IF([test "x${ENABLESYNCD}" = "xno" ],[syncd=false],[])
 AM_CONDITIONAL(SYNCD, test x$syncd = xtrue)
 AM_COND_IF([SYNCD], [
-        AM_COND_IF([SAIVS], [], [AC_CHECK_LIB([sai], [main], [AC_MSG_NOTICE(libsai found)], [AC_MSG_ERROR(libsai is required for syncd)])
+        AM_COND_IF([SAIVPP], [], [AC_CHECK_LIB([sai], [main], [AC_MSG_NOTICE(libsai found)], [AC_MSG_ERROR(libsai is required for syncd)])
 ])])
 
 AC_ARG_ENABLE(rpcserver,
@@ -213,17 +213,12 @@ CXXFLAGS="$SAVED_FLAGS"
 
 AC_SUBST(CXXFLAGS_COMMON)
 
-AC_ARG_WITH(extra-libsai-ldflags,
-[  --with-extra-libsai-ldflags=FLAGS
-                          extra libsai.so flags for vendor library],
-[AC_SUBST(EXTRA_LIBSAI_LDFLAGS, "${withval}")])
-
 AC_SUBST(SAIINC, "-I\$(top_srcdir)/SAI/inc -I\$(top_srcdir)/SAI/experimental -I\$(top_srcdir)/SAI/meta")
 
 AM_COND_IF([SYNCD], [
-AM_COND_IF([SAIVS], [], [
+AM_COND_IF([SAIVPP], [], [
 SAVED_FLAGS="$CXXFLAGS"
-CXXFLAGS="-Xlinker --no-as-needed -lsai -I$srcdir/SAI/inc -I$srcdir/SAI/experimental -I$srcdir/SAI/meta $EXTRA_LIBSAI_LDFLAGS"
+CXXFLAGS="-Xlinker --no-as-needed -lsai -I$srcdir/SAI/inc -I$srcdir/SAI/experimental -I$srcdir/SAI/meta"
 AC_CHECK_FUNCS(sai_query_api_version, [
 AC_MSG_CHECKING([SAI headers API version and library version check])
 AC_TRY_RUN([
@@ -258,8 +253,7 @@ https://github.com/opencomputeproject/SAI/pull/1795 (structs)
 }],
 [AC_MSG_RESULT(ok)],
 [AC_MSG_RESULT(failed)
-AC_MSG_ERROR("SAI headers API version and library version mismatch")])],
-[AC_MSG_ERROR("SAI library libsai.so does not have sai_query_api_version API which is required")])
+AC_MSG_ERROR("SAI headers API version and library version mismatch")])])
 CXXFLAGS="$SAVED_FLAGS"
 ])])
 
@@ -281,9 +275,8 @@ AC_DEFINE([SAI_GIT_REVISION],
 AC_OUTPUT(Makefile
           meta/Makefile
           lib/Makefile
-          vslib/Makefile
+          vpplib/Makefile
           syncd/Makefile
-          syncd/tests/Makefile
           saiplayer/Makefile
           saidump/Makefile
           saisdkdump/Makefile
